ansible/install_ansible.sh

```diff
@@ -14,17 +14,62 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# This step is needed to upgrade ansible to version 2.4.2 which is required for
-# the ceph backend.
-sudo apt-get remove -y ansible
-sudo apt-get purge -y ansible
+# This step is needed to upgrade ansible to version 2.4.x which is required for
+# the ceph backend installation using ceph-ansible.
 
-sudo add-apt-repository ppa:ansible/ansible-2.4
+REQUIRED_ANAIBLE_VER=${REQUIRED_ANAIBLE_VER:-}
 
+SCRIPT_PATH=$(cd $(dirname $0); pwd)
+
+OSDSDOCK_CONFIG=${SCRIPT_PATH}/group_vars/osdsdock.yml
+if ! [ -e ${OSDSDOCK_CONFIG} ]; then
+    echo ceph-ansible config file: ${OSDSDOCK_CONFIG} does not exist. Aborting.
+    exit 1
+fi
+
+# ceph-ansible has ansible version dependencies according to the stable
+# branches we use. See the below page for details.
+# https://docs.ceph.com/ceph-ansible/master/
+
+echo Reading...: ${OSDSDOCK_CONFIG}
+CEPH_ANSIBLE_BRANCH=`grep ceph_ansible_branch: ${OSDSDOCK_CONFIG} | awk '{print $2;}'`
+case ${CEPH_ANSIBLE_BRANCH} in
+"stable-3.0"|"stable-3.1")
+    REQUIRED_ANSIBLE_VER=2.4
+    ;;
+"stable-3.2")
+    REQUIRED_ANSIBLE_VER=2.6
+    ;;
+"stable-4.0")
+    REQUIRED_ANSIBLE_VER=2.8
+    ;;
+*)
+    echo "Unknown ceph-ansible branch: ${CEPH_ANSIBLE_BRANCH}"
+    exit 1
+    ;;
+esac
+
+if [ "`which ansible`" != ""  ]; then
+# ansible installed
+    CURRENT_VER=`ansible --version | egrep ^ansible | awk '{print $2;}' | cut -b -3`
+    if [ "${CURRENT_VER}" != "${REQUIRED_ANSIBLE_VER}" ]; then
+        echo ansible version error
+        echo removing ansible
+        sudo apt-get remove -y ansible
+        sudo apt-get purge -y ansible
+    else
+        # correct version installed. bailout
+        ansible --version
+        exit 0
+    fi
+fi
+
+# incorrect version removed, or no ansible found.
+echo Installing ansible ${REQUIRED_ANSIBLE_VER} required for ceph-ansible ${CEPH_ANSIBLE_BRANCH} branch.
+sudo add-apt-repository -y ppa:ansible/ansible-${REQUIRED_ANSIBLE_VER}
```

Comment:
Is there any effect on the other installation opensds components if we upgrade/downgrade ansible for ceph only?

Reply:
I don't think there would be any effects on other ansible playbook portions of opensds-installer.

Originally, since before I began to work on this topic, 'install_ansible.sh' script has been checking the existing ansible version and uninstalling it if it's not ansible 2.4. This actually comes from only ansible version requirements of https://github.com/ceph/ceph-ansible depending on branches that we use ('stable-3.0')

In other words, we have been happy using ansible 2.4 so far regardless of it's upgrade or downgrade.

In this sense, as I mentioned above ceph-ansible stable-3.1 still requires ansible 2.4 and supports both luminous and mimic. Switching to stable-3.1 is a safe and good idea, IMHO.